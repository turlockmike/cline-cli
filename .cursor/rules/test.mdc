---
description: Rules when writing or running tests
globs: *.test.ts
---
If a snapshot test is failing, you MUST ask the user if the snapshot is correct or it needs to be updated.

When implementing code, follow the TDD approach. Only attempt to fix 1 test at a time. 