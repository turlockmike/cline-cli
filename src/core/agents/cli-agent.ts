import { LanguageModelV1 } from 'ai';
import { createAgent } from '../agent';
import { getEnvironmentInfo, getAgentRules } from '../prompts';
import { TaskHistory } from '../TaskHistory';
import { ALL_TOOLS } from '../tools';
import { getMcpTools } from '../mcp/toolWrapper';
import { ToolManager } from '../../config/ToolManager';

export async function createCLIAgent(model: LanguageModelV1 | Promise<LanguageModelV1>) {
  // Initialize tools
  const tools = { ...ALL_TOOLS };
  
  try {
    // Get MCP tools from tool manager
    const toolManager = new ToolManager();
    await toolManager.initializeDefaults();
    
    // Get tool configurations
    const toolConfigs = await toolManager.listTools();
    
    // Load MCP tools for each configuration
    for (const toolConfig of toolConfigs) {
      try {
        const resolvedConfig = await toolManager.getResolvedToolConfig(toolConfig);
        
        // Convert to the format expected by getMcpTools
        const mcpConfig = {
          mcpServers: {}
        };
        
        for (const server of resolvedConfig.mcpServers) {
          mcpConfig.mcpServers[server.name] = {
            command: server.command,
            args: server.args || [],
            env: server.env || {},
            disabledTools: server.disabledTools
          };
        }
        
        // Get MCP tools for this configuration
        const { tools: mcpTools } = await getMcpTools({ config: mcpConfig });
        
        // Add MCP tools to the agent's tools
        Object.assign(tools, mcpTools);
      } catch (error) {
        console.warn(`Warning: Failed to load MCP tools for configuration '${toolConfig}':`, error);
      }
    }
  } catch (error) {
    console.warn('Warning: Failed to load MCP tools:', error);
  }
  
  // Create and return the agent
  return createAgent({
    name: 'CLI Agent',
    description: 'A helpful CLI agent that can answer questions and perform tasks',
    role: `
${getAgentRules()}
${getEnvironmentInfo()}`,
    model,
    taskHistory: new TaskHistory(),
    tools,
    callSettings: {
    }
  });
} 