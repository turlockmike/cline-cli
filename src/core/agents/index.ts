export { createCLIAgent } from './cli-agent';

// Future agents can be exported here
// export { createWebAgent } from './web-agent';
// export { createChatAgent } from './chat-agent';
// etc... 