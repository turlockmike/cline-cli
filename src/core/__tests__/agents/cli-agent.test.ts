import { MockLanguageModelV1 } from 'ai/test';
import { createCLIAgent } from '../../agents/cli-agent';
import * as os from 'node:os';
import { readdirSync } from 'node:fs';

// Add mock for environment details at the top of the file
jest.mock('../../prompts/environment', () => ({
  getEnvironmentInfo: () => `<environment_details>
Environment Information: static testing info
</environment_details>`
}));

// Mock sound-play
jest.mock('sound-play', () => ({
    play: jest.fn()
}));

// Mock the fs module
jest.mock('node:fs', () => ({
    readdirSync: jest.fn()
}));

// Mock os module
jest.mock('node:os', () => ({
    ...jest.requireActual('node:os'),
    homedir: jest.fn(),
    platform: jest.fn(),
    release: jest.fn(),
    arch: jest.fn(),
    cpus: jest.fn(),
    totalmem: jest.fn(),
    freemem: jest.fn(),
    userInfo: jest.fn()
}));

// Mock child_process
jest.mock('child_process', () => ({
    execSync: jest.fn().mockReturnValue(Buffer.from('HEAD'))
}));

describe('CLI Agent', () => {
    const mockCwd = '/home/runner/work/hataraku/hataraku';
    const mockHomedir = '/home/runner';

    it('should create CLI agent with correct configuration', () => {
        // Mock process values
        process.cwd = jest.fn().mockReturnValue(mockCwd);
        process.env.TERM_PROGRAM = undefined;

        // Mock Date to ensure consistent timestamp
        const mockDate = new Date('2025-02-19T20:18:28.000Z'); // UTC time matching CI
        jest.spyOn(global, 'Date').mockImplementation(() => mockDate);
        
        // Set timezone to UTC
        process.env.TZ = 'UTC';

        // Mock Intl.DateTimeFormat to ensure consistent locale and timezone
        const mockDateTimeFormat = {
            resolvedOptions: () => ({
                locale: 'en-US',
                timeZone: 'UTC',
                calendar: 'gregory',
                numberingSystem: 'latn'
            }),
            format: jest.fn(),
            formatToParts: jest.fn(),
            formatRange: jest.fn(),
            formatRangeToParts: jest.fn()
        };
        global.Intl.DateTimeFormat = Object.assign(jest.fn(() => mockDateTimeFormat), {
            supportedLocalesOf: jest.fn()
        });

        // Setup mock return values to match GitHub Actions environment
        (os.homedir as jest.Mock).mockReturnValue(mockHomedir);
        (os.platform as jest.Mock).mockReturnValue('linux');
        (os.release as jest.Mock).mockReturnValue('6.8.0-1021-azure');
        (os.arch as jest.Mock).mockReturnValue('x64');
        (os.cpus as jest.Mock).mockReturnValue(Array(4).fill({})); // 4 CPUs
        (os.totalmem as jest.Mock).mockReturnValue(17179869184); // 16GB
        (os.freemem as jest.Mock).mockReturnValue(15032385536); // 14GB
        (os.userInfo as jest.Mock).mockReturnValue({ username: 'runner' });
        (readdirSync as jest.Mock).mockReturnValue([
            '.changeset', '.cursor', '.eslintrc.json', '.git', '.gitattributes', 
            '.github', '.gitignore', '.husky', '.npmrc', '.nvmrc', '.prettierignore', 
            '.prettierrc.json', 'CHANGELOG.md', 'CONTRIBUTING.md', 'LICENSE', 'README.md', 
            'docs', 'esbuild.js', 'jest.config.cjs', 'node_modules', 'package-lock.json', 
            'package.json', 'src', 'tsconfig.cjs.json', 'tsconfig.esm.json', 'tsconfig.json'
        ]);

        const mockModel = new MockLanguageModelV1({
            defaultObjectGenerationMode: 'json',
            doGenerate: async () => ({
                text: 'Test response',
                finishReason: 'stop',
                usage: { promptTokens: 10, completionTokens: 20 },
                rawCall: { rawPrompt: null, rawSettings: {} }
            })
        });

        const agent = createCLIAgent(mockModel);
        
        // Create a stable version of the agent for snapshot
        const stableAgent = {
            ...agent,
            modelPromise: expect.any(Promise)
        };
        
        (expect(stableAgent) as any).toMatchSnapshot({
            tools: expect.any(Object),
            taskHistory: {
                historyDir: expect.stringMatching(/^.*\/\.local\/share\/hataraku\/logs$/)
            }
        });

        // Clean up mocks
        jest.restoreAllMocks();
    });
}); 