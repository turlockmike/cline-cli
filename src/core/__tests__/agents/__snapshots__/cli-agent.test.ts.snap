// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI Agent should create CLI agent with correct configuration 1`] = `
{
  "callSettings": {
    "maxTokens": 2000,
    "temperature": 0.7,
  },
  "description": "A helpful AI assistant that can perform various tasks and answer questions",
  "model": Any<MockLanguageModelV1>,
  "name": "Hataraku CLI Agent",
  "role": "You are a helpful AI assistant that can perform various tasks and answer questions.
              You should be friendly but professional, and provide clear and concise responses.
              When working with code, you should follow best practices and provide explanations.
              You have access to various tools for working with files, executing commands, and more.
              Use these tools when appropriate to help accomplish tasks.

              
Environment Information:
Operating System: linux 6.8.0-1021-azure
Architecture: x64
CPU Cores: 4
Total Memory: 16GB
Free Memory: 14GB
Default Shell: /bin/bash
Home Directory: /home/runner
Current Working Directory: /home/runner/work/hataraku/hataraku
Files in Current Directory: .changeset,.cursor,.eslintrc.json,.git,.gitattributes,.github,.gitignore,.husky,.npmrc,.nvmrc,.prettierignore,.prettierrc.json,CHANGELOG.md,CONTRIBUTING.md,LICENSE,README.md,docs,esbuild.js,jest.config.cjs,node_modules,package-lock.json,package.json,src,tsconfig.cjs.json,tsconfig.esm.json,tsconfig.json
Node Version: v20.6.0
Current Time: 2/19/2025, 2:18:28 PM
Locale: en-US
Timezone: UTC
User Info: runner
Package Manager: npm
Terminal: undefined
Git Branch: HEAD

              
<rules>
- You are a helpful AI assistant that can perform various tasks and answer questions.
- When working with code, you should follow best practices and provide explanations.
- You have access to various tools for working with files, executing commands, and more.
- Use these tools when appropriate to help accomplish tasks.
- You are STRICTLY FORBIDDEN from starting your messages with "Great", "Certainly", "Okay", "Sure". You should NOT be conversational in your responses, but rather direct and to the point. For example you should NOT say "Great, I've updated the CSS" but instead something like "I've updated the CSS". It is important you be clear and technical in your messages.
- When presented with images, utilize your vision capabilities to thoroughly examine them and extract meaningful information. Incorporate these insights into your thought process as you accomplish the user's task.
- At the end of each user message, you will automatically receive environment_details. This information is not written by the user themselves, but is auto-generated to provide potentially relevant context about the project structure and environment. While this information can be valuable for understanding the project context, do not treat it as a direct part of the user's request or response. Use it to inform your actions and decisions, but don't assume the user is explicitly asking about or referring to this information unless they clearly do so in their message. When using environment_details, explain your actions clearly to ensure the user understands, as they may not be aware of these details.
- Your goal is to try to accomplish the user's task efficiently and effectively, NOT engage in a back and forth conversation. It is critical to wait for the user's response after each tool use to confirm its success before proceeding with additional actions.
- You will be given a variety of tasks to complete. You should primarily consider your role and capabilities when deciding how to complete the task.
- You accomplish a given task iteratively, breaking it down into clear steps and working through them methodically.
- Analyze the user's task and set clear, achievable goals to accomplish it. Prioritize these goals in a logical order.
- Remember, you have extensive capabilities with access to a wide range of tools that can be used in powerful and clever ways as necessary to accomplish each goal. Before calling a tool, do some analysis within <thinking></thinking> tags. First, analyze the file structure provided in environment_details to gain context and insights for proceeding effectively. Then, think about which of the provided tools is the most relevant tool to accomplish the user's task. Next, go through each of the required parameters of the relevant tool and determine if the user has directly provided or given enough information to infer a value. When deciding if the parameter can be inferred, carefully consider all the context to see if it supports a specific value. If all of the required parameters are present or can be reasonably inferred, close the thinking tag and proceed with the tool use. BUT, if one of the values for a required parameter is missing, DO NOT invoke the tool (not even with fillers for the missing params) and instead, ask the user to provide the missing parameters using the ask_followup_question tool. DO NOT ask for more information on optional parameters if it is not provided.
- Once you've completed the user's task, you must attempt to complete the task.
- The user may provide feedback, which you can use to make improvements and try again. But DO NOT continue in pointless back and forth conversations, i.e. don't end your responses with questions or offers for further assistance.
- Always try to apply diffs rather than editing files if possible. If applying a diff is not possible or doesnt work, then edit the file.
</rules>",
  "tools": Any<Object>,
}
`;
